### PR TITLE
[WIP] [DO NOT REVIEW] Stress test progress notify POC

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

**Comment by jpbetz** [comment]
The `failed to wait for definition ... not to be served anymore: failed to wait for OpenAPI spec validating condition` failures in `pull-kubernetes-e2e-gce` look like they're happening frequently (always?).

CRD openAPI schemas are kept up-to-date using a watch here: https://github.com/kubernetes/kubernetes/blob/15a883b06740ddbb201bd8560524cb22291ce923/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L43

I checked the logs for the `e2e-test-crd-publish-openapi-3091-crd` in run [1212709237235388416](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86769/pull-kubernetes-e2e-gce/1212709237235388416/) and I see:

```
$ egrep "e2e-test-crd-publish-openapi-3091-crd" kube-apiserver.log
...
I0102 12:39:22.080724       1 httplog.go:90] DELETE /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-3091-crds.crd-publish-openapi-test-common-group.example.com: (417.240033ms) 200 ...
...
I0102 12:39:25.198574       1 customresource_discovery_controller.go:283] Deleting customresourcedefinition "e2e-test-crd-publish-openapi-3091-crds.crd-publish-openapi-test-common-group.example.com"
...
I0102 12:39:25.198698       1 controller.go:172] Updating CRD OpenAPI spec because e2e-test-crd-publish-openapi-3091-crds.crd-publish-openapi-test-common-group.example.com was removed
```

Which suggests the openapi spec was successfully removed before the error was reported:

```
Jan  2 12:39:56.606: failed to wait for definition "com.example.crd-publish-openapi-test-common-group.v6.E2e-test-crd-publish-openapi-3091-crd" not to be served anymore: failed to wait for OpenAPI spec validating condition: timed out waiting for the condition; lastMsg: spec.SwaggerProps.Definitions["com.example.crd-publish-openapi-test-common-group.v6.E2e-test-crd-publish-openapi-3091-crd"] still exists
test/e2e/apimachinery/crd_publish_openapi.go:363
```

But I also see:

```
W0102 12:39:30.555257       1 controller.go:142] slow openapi aggregation of "e2e-test-crd-publish-openapi-3091-crds.crd-publish-openapi-test-common-group.example.com": 5.356558033s
```

which might be red herring but seems to suggest openapi aggregation is slower than expected.

Maybe try with frequent progress notify events or with the timeout for openapi waits set much higher here? 
https://github.com/kubernetes/kubernetes/blob/b85adbf1fd9131a153141547fc391221532e1406/test/e2e/apimachinery/crd_publish_openapi.go#L604

**Comment by wojtek-t** [comment]
Thanks @jpbetz !

> Maybe try with frequent progress notify events or with the timeout for openapi waits set much higher here?

It's already very frequent here - every 50ms.
I also tried with 500ms (10x less frequent) and it looked similar.

Let me try retesting one more time and in case of similar failure, i will try with 5s frequency to see if that helps.

**Comment by wojtek-t** [comment]
@jpbetz - so it's clearly related, bumping to 5m fixed all those tests (without any other changes)

**Comment by wojtek-t** [comment]
[I mean, bumping the frequency of progress notify events]

**Comment by wojtek-t** [comment]
The CRD-failures are clearly related to this change
When I bumped the frequency to 5s (from 50ms) the failures went away, so it's clearly related.So I started chasing one of the failures a bit:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86769/pull-kubernetes-e2e-gce/1213059686740267009/
this test in particular
[sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance]

Here are the logs from the test:
```
wojtekt@wojtekt-work:~/Downloads$ cat kube-apiserver.log | grep "/customreso" | grep "renamed" | grep httplog
I0103 12:02:25.845040       1 httplog.go:90] POST /apis/apiextensions.k8s.io/v1/customresourcedefinitions: (147.170701ms) 201 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:56974]
I0103 12:02:26.651827       1 httplog.go:90] GET /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com: (510.394141ms) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:56974]
I0103 12:03:11.238143       1 httplog.go:90] PATCH /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com: (66.159283ms) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36598]
I0103 12:03:46.687959       1 httplog.go:90] DELETE /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com: (256.327054ms) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36724]
```
(that roughly show what was happening when)

And self-requests from apiserver:
```
wojtekt@wojtekt-work:~/Downloads$ cat kube-apiserver.log | grep "multi-ver" | grep "openapi-4957" | grep -v GET | grep "kube-apiserver"
I0103 12:02:26.888680       1 httplog.go:90] PUT /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com/status: (391.583478ms) 200 [kube-apiserver/v1.18.0 (linux/amd64) kubernetes/d206e8f [::1]:60170]
I0103 12:02:27.577794       1 httplog.go:90] PUT /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com/status: (186.049159ms) 200 [kube-apiserver/v1.18.0 (linux/amd64) kubernetes/d206e8f [::1]:60170]
I0103 12:03:47.380121       1 httplog.go:90] PUT /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com/status: (221.900926ms) 200 [kube-apiserver/v1.18.0 (linux/amd64) kubernetes/d206e8f [::1]:60170]
I0103 12:03:48.834933       1 trace.go:116] Trace[182222864]: "Update" url:/apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com/status,user-agent:kube-apiserver/v1.18.0 (linux/amd64) kubernetes/d206e8f,client:::1 (started: 2020-01-03 12:03:47.931311024 +0000 UTC m=+1369.120830306) (total time: 903.591577ms):
I0103 12:03:48.835123       1 httplog.go:90] PUT /apis/apiextensions.k8s.io/v1/customresourcedefinitions/e2e-test-crd-publish-openapi-4957-crds.crd-publish-openapi-test-multi-ver.example.com/status: (903.981889ms) 200 [kube-apiserver/v1.18.0 (linux/amd64) kubernetes/d206e8f [::1]:60170]
```


Now, what is important, the calls to openapi are very frequent, but most of them return 304, if we filter those out, we got:
```
wojtekt@wojtekt-work:~/Downloads$ cat kube-apiserver.log | grep "openapi/v2" | grep renamed | grep -v " 304 "
I0103 12:02:29.827873       1 httplog.go:90] GET /openapi/v2: (723.521312ms) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:56974]
I0103 12:02:47.060207       1 httplog.go:90] GET /openapi/v2: (1.042567614s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36374]
I0103 12:02:54.043090       1 httplog.go:90] GET /openapi/v2: (2.285949643s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36472]
I0103 12:03:01.198692       1 httplog.go:90] GET /openapi/v2: (2.926942953s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36522]
I0103 12:03:06.388894       1 httplog.go:90] GET /openapi/v2: (1.604674983s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36558]
I0103 12:03:13.643004       1 httplog.go:90] GET /openapi/v2: (1.695163855s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36598]
I0103 12:03:31.430661       1 httplog.go:90] GET /openapi/v2: (1.55469634s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36636]
I0103 12:03:38.169481       1 httplog.go:90] GET /openapi/v2: (2.387438591s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36700]
I0103 12:03:42.273434       1 httplog.go:90] GET /openapi/v2: (1.189038033s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36718]
I0103 12:03:51.526101       1 httplog.go:90] GET /openapi/v2: (1.555200032s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36724]
I0103 12:04:07.249835       1 httplog.go:90] GET /openapi/v2: (1.281023062s) 200 [e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] updates the published spec when one version gets renamed [Conformance] 104.197.105.149:36752]
```

Looking into `waitForOpenAPISchema` it seems that we need to wait for first successful 200 call (and we can get 304 as the first response, because client is shared between calls).

IIUC, in this particular test, before we get to PATCH, we need:
- 2 `waitForDefinition` calls for setupCRD (one per version)
- 2 explicit `waitForDefinition` calls

So my personal feeling is that these tests are somewhat broken (though the PR may contribute to the problem).

**Comment by wojtek-t** [comment]
@liggitt - to share his thoughts on this comment: https://github.com/kubernetes/kubernetes/pull/86769#issuecomment-570571160

**Comment by liggitt** [comment]
cc @roycaihw who has been investingating that particular test in https://github.com/kubernetes/kubernetes/issues/86418

**Comment by roycaihw** [comment]
@wojtek-t could you elaborate more on https://github.com/kubernetes/kubernetes/pull/86769#issuecomment-570571160? Are you suggesting the number of 200s is more than expected? Each `waitForOpenAPISchema` call may (actually usually) result in more than one 200

>2 waitForDefinition calls for setupCRD (one per version)
2 explicit waitForDefinition calls

I think the latter 2 explicit calls can be removed as duplicate, but they don't affect the flakiness of the test IMO

**Comment by roycaihw** [comment]
also could you explain what bottleneck does the progress notify interval introduce when it's 50ms?

**Comment by wojtek-t** [comment]
> also could you explain what bottleneck does the progress notify interval introduce when it's 50ms?

Didn't yet fully understand it. But my feeling is that it's purely cpu-starvation - that's generating additional non-negligible load on etcd and apiserver.

**Comment by mm4tt** [comment]
I run 5k node perf tests on this change comparing it to the one of our CI/CD 5k runs (baseline). 

Baseline: http://35.237.122.182:3000/d/KpNdI9EZz/progress-notify-tests?orgId=1&var-source=gce-5k-progress-notify-baseline-1217492047036944384&from=1579107056056&to=1579152996624
This PR: http://35.237.122.182:3000/d/KpNdI9EZz/progress-notify-tests?orgId=1

Both tests passed, but it seems that the progress notify run was slightly worse.
Some relevant metrics:

## api-call latency

**baseline**
vA4WsfkVE47

**progress-notify**
DTbagtdxwn3

The api-call-latency is slightly worse in progress-notify. In progress notify kube-controller-manager restarted once which correlated with big spikes in call latency. Not sure whether this had something to do with the changes introduced in this PR or whether this was a flake.

## etcd compaction pause duration
**baseline**
nsMiWFOjEg4

**progress-notify**
pvMmPWVoQtZ (1)

The compaction pause duration seems to be visible higher in the experimental run.

## CPU usage
**baseline**
tWiCMU4sdv5

**progress-notify**
izYTUvsY4gw

There is a slight increase in CPU usage by both api-server and etcd container in the experimental run. 

## Summary

We should check a few things, e.g. why the compaction pause duration went up. Some things may be expected (e.g. cpu usage) and some might be caused by flakes.
But overall it looks promising. 

I'll re-run the tests to get more data and rule out that some symptoms (e.g. scheduler restart) were flakes.

**Comment by mm4tt** [comment]
**Run #2:**

Grafana: http://35.237.122.182:3000/d/KpNdI9EZz/progress-notify-tests?orgId=1&var-source=gce-5k-progress-notify-2&from=1579257406342&to=1579299822668

The restart of kube-controller-manager turned out to be a flake, but other symptoms seem to be present in both runs, e.g. slightly higher api-call-latency, this is especially visible on PUT lease requests.

**Comment by mm4tt** [comment]
### Run #3 - progress-notify-interval set to 250ms (https://github.com/kubernetes/kubernetes/pull/87520)

**Grafana:** http://35.237.122.182:3000/d/KpNdI9EZz/progress-notify-tests?orgId=1&var-source=gce-5k-progress-notify-250ms&from=1580139979631&to=1580183900833
**Baseline Grafana:** http://35.237.122.182:3000/d/KpNdI9EZz/progress-notify-tests?orgId=1&from=1579971607664&to=1580016033707&var-source=gce-5k-progress-notify-baseline-1221115971889008641

I don't see any significant differences between the baseline and the experiment. All the differences spotted in the previous runs with 50ms interval (e.g. cpu usage, etcd compaction pause duration, api-call-latency) are gone. The only things that stands out is the high api-call-latency of PUT lease requests, but that seems to be occurring in the baseline run as well. 

To summarize, **the performance impact of progress-notify with 250ms interval seems to be negligible**.

**Comment by wojtek-t** [comment]
Closing in favor of #94364
